### PR TITLE
Fix TLS

### LINKED ISSUE #30: TLS only works if webadaptor and server/portal are on the same machine
Webadaptor and Server or Portal can be on different machines. Therefore it will be necessary to Support different certificates for Webadaptor and Server or Portal. Furthermore it could be necessary to Import Root or intermediate Certificates to Portal and Server.

I'll provide PR on this soon.

### PR DESCRIPTION
**Pull request by spitzerr**
Fixes #30

### CODE REVIEW

File: Modules/ArcGIS/DSCResources/ArcGIS_Server_TLS/ArcGIS_Server_TLS.psm1 [hunk 1]
```diff
@@ -1022,12 +1022,12 @@ function Get-SSLCertificateForMachine
                 $null 
             }else {
                 $IssuerValue = $null
-                $Issuer = [regex]::matches($res.Content, '"issuer":"([A-Za-z =,\.0-9\-]+)\"')
+                $Issuer = [regex]::matches($res.Content, '"Issuer":[ ]?\"([A-Za-z =,\.0-9\-]+)\"')
```
Comment (spitzerr) [inline Modules/ArcGIS/DSCResources/ArcGIS_Server_TLS/ArcGIS_Server_TLS.psm1]: On our site `issuer` is all lowercase. Maybe the check could be case-insensitive?

```[1060-303]:                            [[ArcGIS_Server_TLS]Server_TLS_1060-303.gis.lab] POST http://1060-303.gis.lab:6080/arcgis/admin/machines/1060-303.gis.lab/sslcertificates/1060-303.gis.lab with 101-byte payload
[1060-303]:                            [[ArcGIS_Server_TLS]Server_TLS_1060-303.gis.lab] received 641-byte response of content type application/json;charset=UTF-8
[1060-303]:                            [[ArcGIS_Server_TLS]Server_TLS_1060-303.gis.lab] Response {"aliasName":"1060-303.gis.lab","issuer":"CN=gis-lab-CA, DC=gis, DC=lab","subject":"CN=1060-303.gis.lab","subjectAlternativeNames":["DNSName: 1060-303.gis.lab"],"validFrom":"Fri Feb 23 09:32:19 CET 2018","validUntil":"Sun Feb 23 09:32:19 CET 2020","keyAlgorithm":"RSA","keySize":2048,"serialNumber":"31000000191c9486ab18bfe3ee000000000019","version":3,"signatureAlgorithm":"SHA256withRSA","keyUsage":["Digital_Signature","Key_Encipherment"],"md5Fingerprint":"8f09f3f00086ec447dba928856867583","sha1Fingerprint":"6991c2071dddae2974a97c8eb7e2366e089d3f97","sha256Fingerprint":"962c1194a6c1b73eea26bce6138592a15bb11d629d0af4f1d61e2bbd5325d964"}
[1060-303]:                            [[ArcGIS_Server_TLS]Server_TLS_1060-303.gis.lab] Issuer Value:-
[1060-303]:                            [[ArcGIS_Server_TLS]Server_TLS_1060-303.gis.lab] CN:-

File: Modules/ArcGIS/DSCResources/ArcGIS_Server_TLS/ArcGIS_Server_TLS.psm1 [hunk 2]
```diff
@@ -1022,12 +1022,12 @@ function Get-SSLCertificateForMachine
...
                 $Issuer.Groups | %{ 
                     if($_.Value -and $_.Value.Length -gt 0){
-                        $Pos = $_.Value.IndexOf('"issuer"')
+                        $Pos = $_.Value.IndexOf('"Issuer"')
```
Comment (spitzerr) [inline Modules/ArcGIS/DSCResources/ArcGIS_Server_TLS/ArcGIS_Server_TLS.psm1]: On our site `issuer` is all lowercase. Maybe the check could be case-insensitive?

File: Modules/ArcGIS/DSCResources/ArcGIS_Server_TLS/ArcGIS_Server_TLS.psm1 [hunk 3]
```diff
@@ -1022,12 +1022,12 @@ function Get-SSLCertificateForMachine
...
                         if($Pos -gt -1) {
-                            $Str = $_.Value.Substring($Pos + '"issuer"'.Length)
+                            $Str = $_.Value.Substring($Pos + '"Issuer"'.Length)
```
Comment (spitzerr) [inline Modules/ArcGIS/DSCResources/ArcGIS_Server_TLS/ArcGIS_Server_TLS.psm1]: On our site `issuer` is all lowercase. Maybe the check could be case-insensitive?

Comment (shailesh91) [reply Modules/ArcGIS/DSCResources/ArcGIS_Server_TLS/ArcGIS_Server_TLS.psm1]: @spitzerr yes that might be the correct thing to do.

**Comment by spitzerr** [comment]
@shailesh91 fixed the casesensitivity.
ready to merge?

**Comment by shailesh91** [comment]
@spitzerr Not yet. I have to test a lot of things and make a few changes. I have to add fallback in case only the certificates are specified at the server or portal block level.

**Comment by shailesh91** [comment]
@spitzerr Do you have any concerns about the new implementation?

**Comment by spitzerr** [comment]
@shailesh91 you made lots of changes to this. I Need more time to Review this. But I'll give you Feedback as soon as I looked it over

**Comment by shailesh91** [comment]
@spitzerr I agree with your first point. I will make the necessary change. As for the third probably the external DNS name parameter is not being set. Can you tell me what config where you using?
As for the second point, I agree that it is twice the work since 90% of the code base is same. But it is really hard to maintain a single code base for two configurations as their workings are very different. It is likely it will become less accessible and very hard to maintain over time. But since you have brought it up I will consult Nikhil about your concern.

**Comment by spitzerr** [comment]
@shailesh91 I attach my config for this. Maybe you could Elaborate the missing part:
[config.json.txt](https://github.com/Esri/arcgis-powershell-dsc/files/1946024/config.json.txt)
Thank you very much in advance.

File: Modules/ArcGIS/Configuration/ArcGISConfigure.ps1 [hunk 4]
```diff
@@ -904,29 +932,42 @@ Configuration ArcGISConfigure
...
                     $Depends += @('[ArcGIS_Service_Account]Portal_RunAs_Account')
 
-                    $HasLoadBalancer = (($AllNodes | Where-Object { $_.Role -icontains 'LoadBalancer' }  | Measure-Object).Count -gt 0)
-                    $ExternalDNSName = [System.Net.DNS]::GetHostByName($PrimaryPortalMachine).HostName
-                    if($ConfigurationData.ConfigData.Portal.SslCertifcate)
-                    {
-                        $ExternalDNSName = $ConfigurationData.ConfigData.Portal.SslCertifcate.Alias
+                    $ExternalDNSName =(Get-FQDN $PrimaryPortalMachine.NodeName) 
+                    if($Node.SslCertifcate){
+                        $ExternalDNSName = $LBMachine.SslCertifcate.Alias
```
Comment (spitzerr) [inline Modules/ArcGIS/Configuration/ArcGISConfigure.ps1]: @shailesh91  $LBMachine is not defined here. Maybe this Needs to be ```$Node.SslCertifcate.Alias```?

Comment (shailesh91) [reply Modules/ArcGIS/Configuration/ArcGISConfigure.ps1]: @spitzerr Yep.

File: Modules/ArcGIS/Configuration/ArcGISConfigure.ps1 [hunk 5]
```diff
@@ -23,24 +23,30 @@ Configuration ArcGISConfigure
...
-            $PrimaryServerMachine  = $AllNodes[$i].NodeName
+            $PrimaryServerMachineNode = $AllNodes[$i]
+            $PrimaryServerMachine  = $PrimaryServerMachineNode.NodeName
         }
 
         if($Role -icontains 'Portal' -and -not($PrimaryPortalMachine))
         {
-            $PrimaryPortalMachine= $AllNodes[$i].NodeName
+            $PrimaryPortalMachineNode = $AllNodes[$i]
+            $PrimaryPortalMachine= $PrimaryPortalMachineNode.NodeName
```
Comment (spitzerr) [inline Modules/ArcGIS/Configuration/ArcGISConfigure.ps1]: Changing this, breaks other parts of the Code, see below

File: Modules/ArcGIS/Configuration/ArcGISConfigure.ps1 [hunk 6]
```diff
@@ -904,29 +932,42 @@ Configuration ArcGISConfigure
                     
                     $Depends += @('[ArcGIS_Service_Account]Portal_RunAs_Account')
 
-                    $HasLoadBalancer = (($AllNodes | Where-Object { $_.Role -icontains 'LoadBalancer' }  | Measure-Object).Count -gt 0)
-                    $ExternalDNSName = [System.Net.DNS]::GetHostByName($PrimaryPortalMachine).HostName
-                    if($ConfigurationData.ConfigData.Portal.SslCertifcate)
-                    {
-                        $ExternalDNSName = $ConfigurationData.ConfigData.Portal.SslCertifcate.Alias
+                    $ExternalDNSName =(Get-FQDN $PrimaryPortalMachine.NodeName)
```
Comment (spitzerr) [inline Modules/ArcGIS/Configuration/ArcGISConfigure.ps1]: According to the changes above, this Needs to be ```$ExternalDNSName =(Get-FQDN $PrimaryPortalMachine) ``` Didn't it?

**Comment by shailesh91** [comment]
@spitzerr One last thing to do, Add a self-signed cert to Portal Block ([here](https://github.com/spitzerr/arcgis-powershell-dsc/blob/72196679892f26fe0956663789c499cad56b8f91/Modules/ArcGIS/Configuration/ArcGISConfigure.ps1#L1042)) . Once that is done I think we are good to merge

**Comment by spitzerr** [comment]
@shailesh91 The Portal Comes with a Self-Signed cert Out of the box. From my perspective there is nothing to do in this case. Or do I miss something?

**Comment by shailesh91** [comment]
@spitzerr from my understanding, when you install for the first time I don't think so there is a certificate installed. But to be sure, I will confirm this once.

**Comment by spitzerr** [comment]
@shailesh91 if you install server or portal both of them bring certificates (self signed) OoB. Only WebAdaptor needs an handmade cert as it integrates in 3rd party webserver.

**Comment by shailesh91** [comment]
@spitzerr Just confirmed this. The Server and Portal creates a Self Signed Certificate at Service startup time if it is not installed already.

**Comment by shailesh91** [comment]
@spitzerr Everything else looks good. One thing that I saw happening is, the Urls printed out at the end are wrong. Just need to fix those.

**Pull request merged**